src/ufuncs.c

```diff
@@ -1844,6 +1850,71 @@ static void polygonize_func(char** args, npy_intp* dimensions, npy_intp* steps,
 }
 static PyUFuncGenericFunction polygonize_funcs[1] = {&polygonize_func};
 
+static char polygonize_full_dtypes[5] = {NPY_OBJECT, NPY_OBJECT, NPY_OBJECT, NPY_OBJECT,
+                                         NPY_OBJECT};
+static void polygonize_full_func(char** args, npy_intp* dimensions, npy_intp* steps,
+                                 void* data) {
+  GEOSGeometry* geom = NULL;
+  unsigned int n_geoms;
+
+  GEOSGeometry* collection = NULL;
+  GEOSGeometry* cuts = NULL;
+  GEOSGeometry* dangles = NULL;
+  GEOSGeometry* invalidRings = NULL;
+
+  GEOS_INIT;
+
+  GEOSGeometry** geoms = malloc(sizeof(void*) * dimensions[1]);
+  if (geoms == NULL) {
+    errstate = PGERR_NO_MALLOC;
+    goto finish;
+  }
+
+  SINGLE_COREDIM_LOOP_OUTER_NOUT4 {
+    n_geoms = 0;
+    SINGLE_COREDIM_LOOP_INNER {
+      if (!get_geom(*(GeometryObject**)cp1, &geom)) {
+        errstate = PGERR_NOT_A_GEOMETRY;
+        goto finish;
+      }
+      if (geom == NULL) {
+        continue;
+      }
+      geoms[n_geoms] = geom;
+      n_geoms++;
+    }
+    collection =
+        GEOSGeom_createCollection_r(ctx, GEOS_GEOMETRYCOLLECTION, geoms, n_geoms);
+    if (collection == NULL) {
+      errstate = PGERR_GEOS_EXCEPTION;
+      goto finish;
+    }
+
+    GEOSGeometry* ret_ptr =
+        GEOSPolygonize_full_r(ctx, collection, &cuts, &dangles, &invalidRings);
+    if (ret_ptr == NULL) {
+      errstate = PGERR_GEOS_EXCEPTION;
+      goto finish;
+    }
+    OUTPUT_Y_I(1, ret_ptr);
+    OUTPUT_Y_I(2, cuts);
+    OUTPUT_Y_I(3, dangles);
+    OUTPUT_Y_I(4, invalidRings);
+    // GEOSGeom_destroy_r(ctx, collection);
```

Comment:
@caspervdw So I expected that I would need to destroy the collection that is created here (and not returned), but doing so results in segfaults on clean-up / del.

Now, I suppose it is related to this aspect of `GEOSGeom_createCollection`:

> The caller **retains ownership** of the containing array,
> but the ownership of the pointed-to objects is transferred
> to the returned \ref GEOSGeometry.

So the problem here is that `GEOSGeom_destroy_r(ctx, collection)` also destroys the geometries that were put into the collection. But those geometries were the values passed to the ufunc, and thus should not be destroyed. How to solve this?

* Clone all input geometries before putting them in the collection, so the full collection can be destroyed (but that seems wasteful)
* How do I destroy the collection object without destroying the pointed-to objects? Or is that simply not needed (because the collection is nothing else as pointing to those underlying objects)?

Reply:
I think the first option you state is the only viable one. `CreateCollection` transfers ownership of the geometries, so as soon as you use it on geometries which are also owned by a `pygeos.Geometry`, the geometries will have two owners, which is wrong.

Skipping the destroy of the `collection` will lead to a memory leak. So copying is the only way.

Reply:
OK, updated with cloning all input before creating collections